### PR TITLE
(feat) Extension config should be available from the store

### PR DESCRIPTION
**Pull request by brandones**
## Requirements
- [x] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.

#### For changes to apps
- [ ] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide).

#### If applicable
- [x] My work includes tests or is validated by existing tests.
- [ ] I have updated the [esm-framework mock](https://github.com/openmrs/openmrs-esm-core/blob/master/packages/framework/esm-framework/mock.tsx) to reflect any API changes I have made.

## Summary

Per https://github.com/openmrs/openmrs-esm-core/pull/360#issuecomment-1071713166, we are making it possible to access extension config outside the context of the extension itself. This will be useful for allowing config to override meta, which will allow us to create fully generic configurable dashboards.

After this change, we will be able to use code like the following to render dashboards [here](https://github.com/openmrs/openmrs-esm-patient-chart/blob/9bd7a8df9b8ae5757ec90c454571d4da1d4fda5e/packages/esm-patient-chart-app/src/patient-chart/chart-review/chart-review.component.tsx#L32):

```ts
  const ungroupedDashboards = extensionStore.slots['patient-chart-dashboard-slot'].assignedExtensions.map(
    (e) => getDashboardDefinition(e.meta, e.config),
  );
  const groupedDashboards = navGroups
    .map((slotName) => extensionStore.slots[slotName].assignedExtensions.map(
      (e) => getDashboardDefinition(e.meta, e.config))
    ).flat();

function getDashboardDefinition(meta, config) {
  return {  ...meta, ...config };
  // But actually doing something more complicated so that only the expected properties from config
  // get included, etc. The point is, it's up to the extension consumer to implement this logic.
}
```

In order to make this change, I had to refactor the stores so that instead of having a store per extension instance, we have one store that holds all extension configs. This produced most of the churn seen in the PR.

## Other

Re-do of https://github.com/openmrs/openmrs-esm-core/pull/360

### CODE REVIEW

File: packages/framework/esm-extensions/src/render.ts [hunk 1]
```diff
@@ -54,19 +53,27 @@ export function renderExtension(
...
-          [
-            "extensions",
-            extensionName,
-            "instances",
-            extensionSlotModuleName,
-            extensionSlotName,
-          ],
-          { domElement, id: extensionId }
-        )
-      );
```
Comment (brandones) [inline packages/framework/esm-extensions/src/render.ts:69]: Note that in this original implementation, there is only one extension per slot. This is wrong. I don't know why it wasn't producing bugs before, but some other changes I made exposed this bug; hence the change. I suspect that previously, extension configs were never getting updated, so it didn't matter when an extension config disappeared from the store.

File: packages/framework/esm-extensions/src/extensions.ts [hunk 2]
```diff
@@ -271,21 +291,28 @@ export function getConnectedExtensions(
...
+      slotName,
+      id
+    );
     const name = getExtensionNameFromId(id);
     const extension = internalState.extensions[name];
     extensions.push({
       id,
       name,
       moduleName: extension.moduleName,
+      config: extensionConfig,
```
[the remark this answers is not in the document]
Comment (ZacButko) [reply packages/framework/esm-extensions/src/extensions.ts:315]: We want a similar assignment for `meta`, no?

**Review by ZacButko — APPROVED** [review]

File: packages/framework/esm-extensions/src/extensions.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (brandones) [reply packages/framework/esm-extensions/src/extensions.ts:315]: It's on the next line

[the remark this answers is not in the document]
Comment (ZacButko) [reply packages/framework/esm-extensions/src/extensions.ts:315]: Right, but config is getting set with `getExtensionConfigFromStore` and meta is getting set with `internalState.extensions[name].meta` ... I thought the idea was to have both config and meta follow a similar pattern?

[the remark this answers is not in the document]
Comment (brandones) [reply packages/framework/esm-extensions/src/extensions.ts:315]: The meta just comes directly from the extension registration, nothing fancy.

**Pull request merged**